fix(turboself): add fallback balance to prevent card UI block

THE-lulu57 opened the pull request:
![Contribution](https://github.com/PapillonApp/papillon-v8/raw/main/.github/assets/contribution_header.png)

# Règles de contribution
> [!CAUTION]
> Afin de garantir une application stable et pérenne dans le temps, nous t'invitons à vérifier que tu as bien respecté les règles de contribution. Sans cela, ta Pull Request ne pourra pas être examinée.

- [x ] Cette Pull Request porte sur une seule fonctionnalité ou un seul correctif.
- [x ] Cette Pull Request n'est pas faite essentiellement avec de l'IA.
- [ ] Pour tout changement majeur, j’ai créé une issue afin d’échanger avec les mainteneurs de Papillon sur la meilleure façon de l’intégrer.
- [x ] Ma Pull Request respecte les conventions Conventional Commits et Conventional Branch ainsi que les conventions de codage de l'application.
- [ ] J’ai testé mes modifications sur iOS et Android, et l’application fonctionne correctement.
- [x ] J’emploie un langage informel, clair et concis dans mes messages.
- [ x] J’ai documenté mes changements de manière appropriée, soit dans la description de la Pull Request, soit dans le GitBook.
- [ ] J’ai ajouté les traductions nécessaires dans au moins un fichier de langue.

# Résumé des changements

Correction d'un bug qui faisait que, dans certains cas, la carte turboself n'apparaissait juste pas, le login fonctionnait mais aucune fonction n'était appelée ensuite, bloquant l'affichage de la carte.
Pour corriger j'ai : 
- Ajouté "fallbackBalance", un solde de secours en cas d'échec le temps que les infos du compte arrivent à Papillon et se mettent à jour.

# Capture(s) d'écran

<img width="480" height="270" alt="Capture d’écran du 2025-12-23 19-28-21" src="https://github.com/user-attachments/assets/d0195e89-184a-4a96-bf1f-7e120bb30590" />
<img width="518" height="370" alt="Capture d’écran du 2025-12-23 19-28-49" src="https://github.com/user-attachments/assets/ca1f7ff2-3898-4927-b8e4-c7445f83b8e4" />


# Infos en plus : 

J'ai testé les modifs sur Android, n'ayant qu'un seul type de compte, il serait cool que Raphaël ou Maël puissent tester avec les leurs (je crois ils en ont un) pour confirmer que le fallback ne masque pas d'autres soucis de données.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Corrections de bogues**
  * Fiabilité améliorée lors de la récupération des soldes de cantine : validation de session en amont, gestion des erreurs renforcée et utilisation d’un solde de secours pour garantir une réponse stable même en cas d’échec de la récupération.
  * Comportement utilisateur inchangé (signature/API préservée).

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Gabriel29306 commented:
Oh mrd, 3 reviewer pour codeowner, bon bah je reviens plus tard

Gabriel29306 commented:
@Kgeek33 tu pourrais regarder ?

Kgeek33 commented:
> @Kgeek33 tu pourrais regarder ?

ok je regarde ça de suite, ça a l'air intéressant comme pr en lisant la description :)

Gabriel29306 reviewed (commented):
> > @Kgeek33 tu pourrais regarder ?
> 
> ok je regarde ça de suite, ça a l'air intéressant comme pr en lisant la description :) 

Ce serais bien que Pronote fonctionne 😀

Kgeek33 commented:
ok je fais une review de code après mais je remarque déjà des problèmes depuis la maj de Pawnote sur la 7.7.0 : 

1. Quand aucune note prise en compte sur une matière
![1736636353518](https://github.com/user-attachments/assets/942252fc-f2a9-4754-b3f6-82ba324ad797)
![1736636353514](https://github.com/user-attachments/assets/e5566a53-0a52-487a-8dc2-b29e526e7a6a)

2. Les -1/20 prient en compte ! Faudrait genre remplacer par NaN ou supprimer à certains endroits : 
![1736636421074](https://github.com/user-attachments/assets/e3be14de-1a76-40a6-b04e-f32086e0a600)
Du coup, l'impact de la moyenne sur une matière est complètement fausse (pas besoin de capture)

Kgeek33 commented on src/router/navigator/atoms/MenuItem.tsx:
```diff
@@ -1,13 +1,11 @@
 import { Pressable } from "react-native-gesture-handler";
 import * as Haptics from "expo-haptics";
-import Reanimated, { FadeIn, FadeOut, LinearTransition, ZoomIn } from "react-native-reanimated";
+import Reanimated, { FadeIn, FadeOut, LinearTransition } from "react-native-reanimated";
```
protife-en mdr

```suggestion
import Reanimated, {
  FadeIn,
  FadeOut,
  LinearTransition,
} from "react-native-reanimated";
```

Kgeek33 commented on src/router/navigator/atoms/MenuItem.tsx:
```diff
@@ -19,7 +17,7 @@ const MenuItem: React.FC<{
 
   const { options } = descriptor;
-  const label = options.tabBarLabel !== undefined ? options.tabBarLabel : options.title !== undefined ? options.title : route.name;
+  const label = options.tabBarLabel !== undefined ? options.tabBarLabel : options.title ?? route.name;
```
```suggestion
  const label =
    options.tabBarLabel !== undefined
      ? options.tabBarLabel
      : options.title ?? route.name;
```

Kgeek33 commented on src/views/account/Chat/Modals/Chat.tsx:
```diff
@@ -1,30 +1,30 @@
-import { Image, ActivityIndicator, FlatList, ImageBackground, Platform, StyleSheet, Text, TextInput, TouchableOpacity, View, KeyboardAvoidingView,} from "react-native";
-import { useTheme} from "@react-navigation/native";
+import React, { useEffect, useRef, useState } from "react";
+import { Image, ActivityIndicator, FlatList, ImageBackground, Platform, StyleSheet, Text, TextInput, TouchableOpacity, View, KeyboardAvoidingView, } from "react-native";
```
oulala mdrrr


```suggestion
import {
  Image,
  ActivityIndicator,
  FlatList,
  ImageBackground,
  Platform,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
  KeyboardAvoidingView,
} from "react-native";
```

Kgeek33 commented on src/views/account/Chat/Modals/Chat.tsx:
```diff
@@ -1,30 +1,30 @@
 import HTMLView from "react-native-htmlview";
-import Reanimated, {FadeIn, FadeInDown, FadeOut} from "react-native-reanimated";
-import {AccountService} from "@/stores/account/types";
+import Reanimated, { FadeIn, FadeInDown, FadeOut } from "react-native-reanimated";
```
```suggestion
import Reanimated, {
  FadeIn,
  FadeInDown,
  FadeOut,
} from "react-native-reanimated";
```

Kgeek33 commented on src/views/account/Chat/Modals/Chat.tsx:
```diff
@@ -1,30 +1,30 @@
 import getAndOpenFile from "@/utils/files/getAndOpenFile";
-import {getProfileColorByName} from "@/services/local/default-personalization";
+import { getProfileColorByName } from "@/services/local/default-personalization";
 import { getChatMessages, sendMessageInChat, getChatRecipients } from "@/services/chats";
```
pas nécessaire mais je propose quand même mdr

```suggestion
import {
  getChatMessages,
  sendMessageInChat,
  getChatRecipients,
} from "@/services/chats";
```

Kgeek33 reviewed (commented):
Voilà, des p'tites optimisations de code qu'on peut faire

Gabriel29306 commented:
@Kgeek33 c'est corrigé
##### J'ai pas corrigé pour les imports, ça peut se faire dans une pr à part

Kgeek33 commented:
Je review plus tard

Kgeek33 commented:
ah je viens de comprendre pourquoi l'influence de la moyenne est complètement fausse… 

ça filtre qu'une note et non toutes les notes d'une matière 🤦‍♂️
https://github.com/Gabriel29306/PapillonV7/blob/75f455f69f1c91de0aa603f3b11e093f672189f1/src/utils/grades/getAverages.ts#L167

mais j'suis en train de review et j'vais te proposer un fichier bien clean

Kgeek33 commented on src/utils/grades/getAverages.ts:
- Ligne 166 : 
```const baseWithoutGradeAverage = getSubjectAverage(
      list.filter((grade) => JSON.stringify(grades[0]) !== JSON.stringify(grade)),
      target,
      useMath
    );```

changer la logique de tri

- Ligne 173 : `difference: baseWithoutGradeAverage - baseAverage, // Calculer la différence entre les deux `

ce n'est pas l'inverse ?

- Ligne 74 : 
```let calcGradesSum = 0; // Somme cumulée des notes pondérées
    let calcOutOfSum = 0; // Somme cumulée des coefficients pondérés

    let countedGrades = 0;```

c'est là où j'ai vrmt été bloqué, à chaque `continue`, ces variables sont réinitialisées à 0, donc fausse tout.

Kgeek33 reviewed (commented):
Depuis 4 heures, j'ai cherché à comprendre mais pas réussi
Si tu trouves, bravo d'avance !

Gabriel29306 commented:
> Depuis 4 heures, j'ai cherché à comprendre mais pas réussi
> Si tu trouves, bravo d'avance !

Je regarderai peut-être ça demain, je boss à implémenter les bulletins dans Pawnote. 
Après ça peut être amusant de trouver le bug

Kgeek33 commented:
> > Depuis 4 heures, j'ai cherché à comprendre mais pas réussi
> > Si tu trouves, bravo d'avance !
> 
> Je regarderai peut-être ça demain, je boss à implémenter les bulletins dans Pawnote. Après ça peut être amusant de trouver le bug

ok c'est cool ça comme nouveauté, perso les notifications (voir #606)
oui, j'avais beau regardé tout le code, ça se remettait à 0 mais en mettant les let en dehors de la fonction, ça fonctionne mais quand on reva sur la même page, ça conserve l'ancienne valeur donc ça fausse encore les informations...

ecnivtwelve commented:
Que fait précisément cette PR?

Gabriel29306 commented:
Faut que je modifie un truc plutard, j'ai fait une petite bêtise dedans

Gabriel29306 reviewed (commented):
Bon, @ecnivtwelve .
Alors, cette PR elle:
- met à jour Pawnote à sa dernière version (`1.4.0`)
- change sur le document d'une note: `-1.00` => `??`
- résout des problèmes de lint (surtout des imports inutilisés)
- aide au débuggage en améliorant un peu les erreurs quand une partie de service n'est pas mise à jour
- dégage les notes comptées comme `-1` (N.Not) dans l'historique des notes.
- supprime un log de debug dans `src/views/settings/SettingsAbout.tsx`

Et voilà

Kgeek33 commented:
tu peux me redemander ma review stp pour que je pense à retester stp

Gabriel29306 reviewed (commented):
@Kgeek33

Kgeek33 commented:
> @Kgeek33

je fais un tour sur Parcoursup et je te review ça

Gabriel29306 reviewed (commented):
Je vais râler sur celui qui a pété mon lint

Gabriel29306 reviewed (commented):
> Je vais râler sur celui qui a pété mon lint 

Tryon je te vois

Kgeek33 commented:
J'suis juste en train d'arranger le `getAverage.ts`, mais LGTM

Kgeek33 commented on src/utils/grades/getAverages.ts:
fais le code suivant :

```
// On importe le type `Grade` depuis le chemin spécifié
import type { Grade } from "@/services/shared/Grade";

// Définition de l'interface `GradeHistory` pour représenter l'historique des notes avec une valeur numérique et une date
export interface GradeHistory {
  value: number; // La valeur de la note
  date: string; // La date à laquelle la note a été enregistrée
}

// Définition du type `Target` qui indique quel type de moyenne ou note cibler
type Target = "student" | "average" | "min" | "max";

// Définition du type `AverageDiffGrade` pour calculer la différence entre les moyennes avec et sans certaines notes
export type AverageDiffGrade = {
  difference?: number; // La différence de moyenne entre deux ensembles de notes
  with: number; // La moyenne avec toutes les notes
  without: number; // La moyenne sans certaines notes
};

// Fonction pour calculer la moyenne des notes globales par matière, en fonction de la cible (par défaut, "student")
const getPronoteAverage = (
  grades: Grade[],
  target: Target = "student",
  useMath: boolean = false
): number => {
  try {
    // Si aucune note n'est fournie ou que la liste est vide, on retourne -1
    if (!grades || grades.length === 0) return -1;

    // Grouper les notes par matière
    const groupedBySubject = grades.reduce(
      (acc: Record<string, Grade[]>, grade) => {
        (acc[grade.subjectId || grade.subjectName] ||= []).push(grade); // Ajouter la note à la liste des notes pour la matière correspondante
        return acc;
      },
      {}
    );

    let countedSubjects = 0;

    // Calculer la moyenne totale de toutes les matières
    const totalAverage = Object.values(groupedBySubject).reduce(
      (acc, subjectGrades) => {
        const nAvg = getSubjectAverage(subjectGrades, target, useMath);

        if (nAvg !== -1) {
          countedSubjects++;
        } else {
          return acc;
        }

        return acc + nAvg;
      },
      0
    );

    // Retourner la moyenne globale en divisant par le nombre de matières
    return totalAverage / countedSubjects;
  } catch (e) {
    return -1;
  }
};

// Fonction pour calculer la moyenne d'une matière spécifique, selon la cible choisie
export const getSubjectAverage = (
  subject: Grade[],
  target: Target = "student",
  useMath: boolean = false,
  loop: boolean = false
): number => {
  try {
    let calcGradesSum = 0; // Somme cumulée des notes pondérées
    let calcOutOfSum = 0; // Somme cumulée des coefficients pondérés
    let countedGrades = 0;

    console.log("début du for");
    for (const grade of subject) {
      console.log("old mg =>", (calcGradesSum / calcOutOfSum) * 20);
      const targetGrade = grade[target];

      if (
        !targetGrade ||
        targetGrade.disabled ||
        targetGrade.value === null ||
        targetGrade.value < 0 ||
        grade.coefficient === 0 ||
        typeof targetGrade.value !== "number"
      ) {
        console.log("t foutu");
        continue;
      }

      const coefficient = grade.coefficient;
      const outOfValue = grade.outOf.value!;
      console.log(
        "coefficient =>",
        coefficient,
        ", note =>",
        targetGrade.value,
        ", noté sur =>",
        outOfValue
      );

      if (grade.isOptional && !loop) {
        const avgWithout = getSubjectAverage(
          subject.filter((g) => JSON.stringify(g) !== JSON.stringify(grade)),
          target,
          useMath,
          true
        );

        const avgWith = getSubjectAverage(subject, target, useMath, true);

        if (avgWithout > avgWith) {
          continue;
        }
      }

      if (grade.isBonus) {
        const averageMoy = outOfValue / 2;
        const newGradeValue = targetGrade.value - averageMoy;

        if (newGradeValue < 0) continue;

        calcGradesSum += newGradeValue;
        calcOutOfSum += 1;
      } else if (useMath) {
        calcGradesSum += targetGrade.value * coefficient;
      } else if (
        targetGrade.value > 20 ||
        (coefficient < 1 && outOfValue - 20 >= -5) ||
        outOfValue > 20
      ) {
        const gradeOn20 = (targetGrade.value / outOfValue) * 20;
        calcGradesSum += gradeOn20 * coefficient;
        calcOutOfSum += 20 * coefficient;
      } else {
        calcGradesSum += targetGrade.value * coefficient;
        calcOutOfSum += outOfValue * coefficient;
      }

      countedGrades += useMath ? 1 * coefficient : 1;
      console.log("new mg =>", (calcGradesSum / calcOutOfSum) * 20);
    }
    console.log("fin du for");

    if (useMath) {
      return calcGradesSum / countedGrades;
    }

    if (calcOutOfSum === 0) return -1;

    const subjectAverage = Math.min((calcGradesSum / calcOutOfSum) * 20, 20);
    return isNaN(subjectAverage) ? -1 : subjectAverage;
  } catch (e) {
    return -1;
  }
};

// Fonction pour calculer la différence de moyenne avec et sans certaines notes
const getAverageDiffGrade = (
  grades: Grade[],
  list: Grade[],
  target: Target = "student",
  useMath: boolean = false
): AverageDiffGrade => {
  try {
    const baseAverage = getSubjectAverage(list, target); // Calculer la moyenne de base avec toutes les notes
    const baseWithoutGradeAverage = getSubjectAverage(
      list.filter(
        (grade) =>
          !grades.some(
            (g) =>
              g.student.value === grade.student.value &&
              g.coefficient === grade.coefficient
          )
      ),
      target,
      useMath
    ); // Calculer la moyenne sans toutes les notes de `grades`

    return {
      difference: baseWithoutGradeAverage - baseAverage, // Calculer la différence entre les deux moyennes
      with: baseAverage, // Moyenne avec toutes les notes
      without: baseWithoutGradeAverage, // Moyenne sans certaines notes
    };
  } catch (e) {
    return {
      difference: 0,
      with: 0,
      without: 0,
    };
  }
};

// Fonction pour générer un historique des moyennes au fil du temps
const getAveragesHistory = (
  grades: Grade[],
  target: Target = "student",
  final?: number,
  useMath: boolean = false
): GradeHistory[] => {
  try {
    // Générer l'historique des moyennes jusqu'à la date de chaque note
    const history = grades.map((grade, index) => ({
      value: getPronoteAverage(grades.slice(0, index + 1), target), // Moyenne jusqu'à ce point
      date: new Date(grade.timestamp).toISOString(), // Date de la note au format ISO
    }));

    // Trier l'historique par date
    history.sort((a, b) => a.date.localeCompare(b.date));

    // Ajouter un point final avec la moyenne finale (ou calculée)
    history.push({
      value: final ?? getPronoteAverage(grades, target, useMath), // Moyenne finale ou calculée
      date: new Date().toISOString(), // Date actuelle
    });

    // remove NaN values
    return history.filter((x) => !isNaN(x.value) || x.value !== -1);
  } catch (e) {
    return [];
  }
};

// Exportation des fonctions pour utilisation externe
export { getPronoteAverage, getAverageDiffGrade, getAveragesHistory };
```

Gabriel29306 replied:
Tu es sûr pour le `console.log("t foutu");` ?

Gabriel29306 replied:
Je vais retirer les logs de débug

Kgeek33 replied:
ah oui mince, j'ai oublié de supprimer les `console.log` 😅
cherche pas pour le t foutu mdrrr

Kgeek33 reviewed (changes requested):
juste ça et c'est ok !

Gabriel29306 commented:
PS: J'ai galéré à trouvé ton ID pour te mettre en tant que Co-authored

Kgeek33 commented:
> PS: J'ai galéré à trouvé ton ID pour te mettre en tant que Co-authored

Ah mince 😅, c'est mon "𝕂𝕪𝕝𝕚𝕒𝕟" qui va pas ?
J'crois que ça a pas marché sur ton dernier commit

Kgeek33 reviewed (approved):
Okay c'est tout bon pour moi
Ça résout un bug sur l'influence d'une matière sur la moyenne générale, mais jsplus quelle issue signale ça

Gabriel29306 commented:
> LOG [ERROR][2025-01-19T17:32:17.886Z][UNKOWN] grades not updated, see:TypeError: Cannot read property 'V' of undefined
> 
> Une idée ? Je peux me plonger dans le code, mais si tu vois d'ou ça peut venir... Aucune note n'est load du coup

Après avoir copié mon code, il faut faire `npm ci` pour réinstaller les dépendances avec les bonnes versions

Gabriel29306 commented:
> > > LOG [ERROR][2025-01-19T17:32:17.886Z][UNKOWN] grades not updated, see:TypeError: Cannot read property 'V' of undefined
> > > Une idée ? Je peux me plonger dans le code, mais si tu vois d'ou ça peut venir... Aucune note n'est load du coup
> > 
> > 
> > Après avoir copié mon code, il faut faire `npm ci` pour réinstaller les dépendances avec les bonnes versions
> 
> J'ai supprimé le dossier `node_modules` et fais `npm ci`, toujours la même erreur...

Fait `npm list` et regarde la version de pawnote

Gabriel29306 commented:
Ah, c'est po normal ça

Gabriel29306 commented:
Enfin, le lint est passé !

Gabriel29306 commented:
Comparé à ce qui a précédemment été fait, maintenant, au lieu d'afficher `??/20` ce n'est tout simplement pas affiché

Gabriel29306 commented:
> Toujours le même problème... `LOG [ERROR][2025-01-25T18:11:27.823Z][UNKOWN] grades not updated, see:TypeError: Cannot read property 'V' of undefined` Et je suis bien sur Pawnote 1.4.0

Je vois pas comment tu peux avoir cette erreur, vraiment.
Tu as vidé ton cache de build (normlement il n'y a pas besoin, mais vu ton cas...) ?

Gabriel29306 commented:
À chaque résolution de conflits, je me mange des nouvelles erreurs TS, youpi

Gabriel29306 commented:
> `npm ci` + `npx expo start -c` et déco / reco du compte, et toujours le même message...

Faudrait que tu ouvres une issue du côté de Literate

Kgeek33 commented:
@Gabriel29306 
Toutes les personnes qui ont signalé des problèmes après le problème de Pawnote rencontrent la même erreur que @Bulgus et ça c'est depuis la maj de Pronote

C'est pas nouveau mais oui, il faut ouvrir une issue depuis Pawnote et non Papillon (pour le moment :))

Gabriel29306 commented:
> @Gabriel29306 
> Toutes les personnes qui ont signalé des problèmes après le problème de Pawnote rencontrent la même erreur que @Bulgus et ça c'est depuis la maj de Pronote
> 
> C'est pas nouveau mais oui, il faut ouvrir une issue depuis Pawnote et non Papillon (pour le moment :)) 

Mais chez moi ça marche, j'ai fait la PR chez Pawnote et elle a été merge

Kgeek33 commented:
> > @Gabriel29306 
> > Toutes les personnes qui ont signalé des problèmes après le problème de Pawnote rencontrent la même erreur que @Bulgus et ça c'est depuis la maj de Pronote
> > 
> > C'est pas nouveau mais oui, il faut ouvrir une issue depuis Pawnote et non Papillon (pour le moment :)) 
> 
> Mais chez moi ça marche, j'ai fait la PR chez Pawnote et elle a été merge

Oui ça marche de mon côté également mais sur certains établissements, il doit y avoir une subtilité qui fait que cette erreur survient
Ça peut être les moyennes générales désactivées tout comme d'autres choses...

Gabriel29306 commented:
> Oui ça marche de mon côté également mais sur certains établissements, il doit y avoir une subtilité qui fait que cette erreur survient
> Ça peut être les moyennes générales désactivées tout comme d'autres choses...

Ils font suer, mais on continuera à s'adapter, pas le choix

Kgeek33 commented:
Faudrait que j'intègre `ncu` pour mettre à jour vers les dernières dépendances (sans toucher à expo 51 => 52)

Gabriel29306 commented:
C'est bon, j'ai mis à jour Pawnote ici

Gabriel29306 commented:
Et je me suis (encore) occupé du lint

Gabriel29306 commented:
<@557981338439778304>

Gabriel29306 commented:
Je vais m'énerver sur un fucking merge car j'ai dû faire un minimum de lint ????

Gabriel29306 commented:
Fait suer

THE-lulu57 committed:
fix: turboself card hidden

Added a backup balance to allow time for information to arrive; otherwise, none of the following functions would be called.

THE-lulu57 committed:
remove spaces

THE-lulu57 committed:
add "error("Session is not valid", "TurboSelf.getCanteenBalances")" like before ;

THE-lulu57 committed:
Add error logging for balance fetch failure

Log an error message when fetching balances fails.

THE-lulu57 committed:
Merge branch 'PapillonApp:main' into main
